adala/utils/llm.py

```diff
@@ -0,0 +1,235 @@
+import asyncio
+import instructor
+import litellm
+import traceback
+import multiprocessing as mp
+from typing import Optional, Dict, Any, List, Type, Union
+from pydantic import BaseModel, Field
+
+instructor_client = instructor.from_litellm(litellm.completion)
+async_instructor_client = instructor.from_litellm(litellm.acompletion)
+
+
+class LLMResponse(BaseModel):
+    data: Dict = Field(default_factory=dict)
+    adala_error: bool = Field(default=False, serialization_alias='_adala_error')
+    adala_message: str = Field(default=None, serialization_alias='_adala_message')
+    adala_details: str = Field(default=None, serialization_alias='_adala_details')
+
+
+def get_messages(user_prompt: str, system_prompt: Optional[str] = None, instruction_first: bool = True):
+    messages = [{'role': 'user', 'content': user_prompt}]
+    if system_prompt:
+        if instruction_first:
+            messages.insert(0, {'role': 'system', 'content': system_prompt})
+        else:
+            messages[0]['content'] += system_prompt
+    return messages
+
+
+async def async_get_llm_response(
+    user_prompt: Optional[str] = None,
+    system_prompt: Optional[str] = None,
+    messages: Optional[List[Dict[str, str]]] = None,
+    model: Optional[str] = 'gpt-4o-mini',
+    instruction_first: Optional[bool] = True,
+    response_model: Optional[Type[BaseModel]] = None,
+    api_key: Optional[str] = None,
+    max_tokens: Optional[int] = 1000,
+    temperature: Optional[float] = 0.0,
+    timeout: Optional[Union[float, int]] = None,
+) -> Dict[str, Any]:
+    """
+    Async version of create_completion function with error handling and session timeout.
+
+    Args:
+        model: model name. Refer to litellm supported models for how to pass
+               this: https://litellm.vercel.app/docs/providers
+        user_prompt: User prompt.
+        system_prompt: System prompt.
+        messages: List of messages to be sent to the model. If provided, `user_prompt`, `system_prompt` and `instruction_first` will be ignored.
+        api_key: API key, optional. If provided, will be used to authenticate
+                 with the provider of your specified model.
+        instruction_first: Whether to put instructions first.
+        response_model: Pydantic model to constrain the LLM generated response. If not provided, the raw completion text will be returned.  # noqa
+        max_tokens: Maximum tokens to generate.
+        temperature: Temperature for sampling.
+        timeout: Timeout in seconds.
+
+    Returns:
+        Dict[str, Any]: OpenAI response or error message.
+    """
+
+    if not user_prompt and not messages:
```

Comment:
This api feels clunky, why allow both "user_prompt + system_prompt" and "messages"? Also doesn't seem we use (or test) passing messages directly anyway

Reply:
I had to leave this for compatibility with `agent.learn()` calls, for example: https://github.com/HumanSignal/Adala/blob/master/adala/skills/_base.py#L318

Reply:
made a TODO to deprecate this "raw" call entirely, removing it from `learn()`